File: python/src/v3/api_object.py
import requests

from api_common import ApiCommon
from utils import input_one_of

class PagedIterator:
    """
    This class implements paging on top of the bookmark functionality provided
    by the Pinterest API class. It hides the paging mechanism behind an iterator.
    """
    def _get_response(self, path_maybe_with_bookmark):
        """
        Use api_object to run HTTP GET. Then, look for bookmark in response.
        """
        response = self.api_object.get_response(path_maybe_with_bookmark)
        unpacked = self.api_object.unpack(response)
        self.data = unpacked.get('data')
        self.bookmark = unpacked.get('bookmark')
        self.index = 0

    def __init__(self, api_object, path):
        """
        Save the api_object and path for subsequent pages of information.
        """
        self.api_object = api_object
        self.path = path # to be used with the bookmark on subsequent requests
        self._get_response(path) # first time, get response without bookmark

    def __iter__(self):
        return self

    def __next__(self):
        if self.index >= len(self.data):
            # need to fetch more data, if there is a bookmark
            if self.bookmark:
                # Determine whether the query needs to be added to the path or
                # if the bookmark will be an additional parameter at the end of the query.
                delimiter = '&' if '?' in self.path else '?'
                path_with_bookmark = self.path + delimiter + 'bookmark=' + self.bookmark
                self._get_response(path_with_bookmark)
                if not self.data: # in case there is some sort of error
                    raise StopIteration
            else:
                raise StopIteration # no bookmark => all done

        retval = self.data[self.index] # get the current element
        self.index += 1 # increment the index for the next time
        return retval

class ApiObject(ApiCommon):
    def __init__(self, api_config, access_token):
        self.api_uri = api_config.api_uri
        self.api_config = api_config
        self.access_token = access_token

    def get_response(self, path):
        if self.api_config.verbosity >= 2:
            print(f'GET {self.api_uri + path}')
        return requests.get(self.api_uri + path, headers=self.access_token.header(), allow_redirects=False)

    def request_data(self, path):
        return self.unpack(self.get_response(path)).get('data')

    def put_data(self, path, put_data):
        if self.api_config.verbosity >= 2:
            print(f'PUT {self.api_uri + path}')
        if self.api_config.verbosity >= 3:
            print(put_data)
        response = requests.put(self.api_uri + path, data = put_data, headers=self.access_token.header(), allow_redirects=False)
        return self.unpack(response).get('data')

    def post_data(self, path):
        if self.api_config.verbosity >= 2:
            print(f'POST {self.api_uri + path}')
        response = requests.post(self.api_uri + path, headers=self.access_token.header(), allow_redirects=False)
        return self.unpack(response).get('data')

    def get_iterator(self, path):
        return PagedIterator(self, path)

    @classmethod
    def print_multiple(klass, page_size, object_name, object_class, paged_iterator):
        """
        Use the PagedIterator to print multiple objects.
        """
        index = 1
        page_index = 1
        for object_data in paged_iterator:
            # do this check after fetching a new page to make sure that there are more pins
            if page_index > page_size:
                if 'yes' == input_one_of(f'Continue printing {object_name} list?', ['yes', 'no'], 'yes'):
                    page_index = 1
                else:
                    break

            # print the object
            print(f'[{index}] ', end='')
            object_class.print_summary(object_data)

            # increment counters
            index += 1
            page_index += 1

File: nodejs/src/v3/api_object.js
import got from 'got'
import {ApiCommon} from '../api_common.js'
import {Input} from '../utils.js'

/**
 * The ApiObject uses the got library for REST transations:
 *   https://www.npmjs.com/package/got
 */
export class ApiObject extends ApiCommon {
  constructor(api_config, access_token) {
    super();
    this.api_uri = api_config.api_uri;
    this.api_config = api_config;
    this.access_token = access_token;
  }

  // Code that is common to a simple GET as in response_data()
  // or to a bookmarked GET as in get_iterator().
  async get_response(path) {
    const full_uri = this.api_uri + path;
    if (this.api_config.verbosity >= 2) {
      console.log('GET', full_uri);
    }
    try {
      const response = await got.get(full_uri, {
        headers: this.access_token.header(),
        followRedirect: false,
        responseType: 'json'
      });
      this.print_response(response);
      return response.body; // success
    } catch (error) {
      this.print_and_throw_error(error);
    }
  }

  // Simple GET transaction.
  async request_data(path) {
    const response = await this.get_response(path);
    return(response.data);
  }

  // Simple PUT transaction.
  async put_data(path, put_data) {
    const full_uri = this.api_uri + path;
    if (this.api_config.verbosity >= 2) {
      console.log('PUT', full_uri);
    }
    if (this.api_config.verbosity >= 3) {
      console.log(put_data);
    }
    try {
      const response = await got.put(full_uri, {
        headers: this.access_token.header(),
        json: put_data,
        followRedirect: false,
        responseType: 'json'
      });
      this.print_response(response);
      return response.body.data;
    } catch (error) {
      this.print_and_throw_error(error);
    }
  }

  async delete_and_check(path) {
    if (this.api_config.verbosity >= 2) {
      console.log('DELETE', this.api_uri + path);
    }
    try {
      const response = await got.delete(this.api_uri + path, {
        headers: this.access_token.header(),
        followRedirect: false
      });
    } catch (error) {
      this.print_and_throw_error(error);
    }
  }

  /**
   *  The iterator returned by this function implements paging on top of the bookmark
   *  functionality provided by the Pinterest API. Useful blog on async iterators:
   *    https://blog.risingstack.com/async-iterators-in-node-js/
   */
  get_iterator(path) {
    const api_object = this;
    return {
      [Symbol.asyncIterator]: async function*() {
        var path_with_query = null; // may not need to be initialized

        // first time: no bookmark
        var response = await api_object.get_response(path);

        // extend the path if there is a bookmark
        if (response.bookmark) {
          path_with_query = path + ((path.includes('?')) ? '&' : '?') + 'bookmark=';
        }

        while (true) {
          // send the current response to the function using the iterator
          for (const value of response.data) {
            yield value;
          }
          // continue the loop if the current response has a bookmark, otherwise done
          if (response.bookmark) {
            response = await api_object.get_response(path_with_query + response.bookmark);
          } else {
            break;
          }
        }
      }
    }
  }

  /* Use the paged_iterator to print multiple objects. */
  async print_multiple(page_size, object_name, object_class, paged_iterator) {
    const input = new Input();
    var index = 1;
    var page_index = 1;
    try {
      for await (let object_data of paged_iterator) {
        // do this check after fetching a new page to make sure that there are more pins
        if (page_index > page_size) {
          if ('yes' == await input.one_of(`Continue printing ${object_name} list?`,
                                          ['yes', 'no'], 'yes')) {
            page_index = 1;
          } else {
            break;
          }
        }
        // print the object
        process.stdout.write(`[${index}] `);
        object_class.print_summary(object_data);

        // increment counters
        index++;
        page_index++;
      }
    } finally {
      input.close();
    }
  }
}
